File: quant1x/core/decode.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import yaml
from typing import Any
from dataclasses import is_dataclass, fields
from .defaults import apply_defaults


def decode_to(dst_type_or_obj: Any, src: Any) -> Any:
    """Decode src (usually a dict) into dst_type_or_obj.

    - If dst_type_or_obj is a type (class), we construct an instance and populate attributes from src.
    - If dst_type_or_obj is an instance, we set attributes on it and return it.
    - For basic target types (int/str/bool/float), we attempt conversion and return the value.
    """
    # basic scalar target conversion
    basic_types = (int, str, bool, float)
    if isinstance(dst_type_or_obj, type) and dst_type_or_obj in basic_types:
        try:
            return dst_type_or_obj(src)
        except Exception as e:
            raise ValueError(f"cannot convert {src!r} to {dst_type_or_obj}: {e}")

    # prepare destination instance
    if isinstance(dst_type_or_obj, type):
        dst = dst_type_or_obj()
        # ensure defaults are present before applying config values
        try:
            apply_defaults(dst)
        except Exception:
            pass
    else:
        dst = dst_type_or_obj

    # if src is not a dict and dst is a simple object, attempt conversion via yaml bridge
    if not isinstance(src, dict):
        dumped = yaml.safe_dump(src)
        loaded = yaml.safe_load(dumped)
        if not hasattr(dst, "__dict__"):
            return loaded
        src = loaded

    # merge dict src into dst instance attributes, skipping None values
    if isinstance(src, dict) and hasattr(dst, "__dict__"):
        for k, v in src.items():
            # skip explicit nulls to preserve defaults
            if v is None:
                continue

            # if existing attribute is a dataclass and incoming value is a dict, merge recursively
            try:
                if hasattr(dst, k):
                    cur = getattr(dst, k)
                else:
                    cur = None
                if cur is not None and is_dataclass(cur) and isinstance(v, dict):
                    decode_to(cur, v)
                    continue

                # dict merger: update existing dict rather than replace entirely
                if isinstance(cur, dict) and isinstance(v, dict):
                    for kk, vv in v.items():
                        if vv is None:
                            continue
                        cur[kk] = vv
                    continue

                # otherwise set attribute (perform basic conversion via yaml bridge)
                try:
                    # attempt to convert scalars via yaml to respect booleans/numbers
                    dumped = yaml.safe_dump(v)
                    loaded = yaml.safe_load(dumped)
                    setattr(dst, k, loaded)
                except Exception:
                    try:
                        setattr(dst, k, v)
                    except Exception:
                        pass
            except Exception:
                # ignore unknown/read-only attributes
                pass
        return dst

    # fallback: use yaml bridge to construct a plain value
    dumped = yaml.safe_dump(src)
    return yaml.safe_load(dumped)


def lookup_config(path: str, root: dict) -> tuple[Any, bool]:
    """Lookup dot-separated path inside root dict. Returns (value, True) or (None, False)."""
    if path == "":
        return root, True
    parts = [p.strip() for p in path.split(".") if p.strip()]
    cur = root
    for p in parts:
        if isinstance(cur, dict) and p in cur:
            cur = cur[p]
        else:
            return None, False
    return cur, True


def decode_config(path: str, dst_type_or_obj: Any, root: dict) -> Any:
    val, ok = lookup_config(path, root)
    if not ok:
        raise KeyError(f"path not found: {path}")
    # create/prepare destination instance first and apply defaults
    if isinstance(dst_type_or_obj, type):
        obj = dst_type_or_obj()
    else:
        obj = dst_type_or_obj

    try:
        apply_defaults(obj)
    except Exception:
        pass

    # then merge configuration values into the object, only overwriting when present
    result = decode_to(obj, val)
    return result

File: quant1x/core/decode.cpp
#include "decode.h"
#include "defaults.h"
#include <yaml-cpp/yaml.h>
#include <sstream>

namespace quant1x {
namespace core {

// 辅助函数: 递归查找map中的路径
YAML::Node lookup_map(const YAML::Node& node, const std::string& path) {
    std::istringstream iss(path);
    std::string token;
    YAML::Node current = node;
    while (std::getline(iss, token, '.')) {
        if (current.IsMap() && current[token]) {
            current = current[token];
        } else {
            return YAML::Node(); // 返回空节点表示未找到
        }
    }
    return current;
}

// LookupConfig 从配置map中按路径查找值
std::any lookup_config(const std::string& path) {
    const auto& config_map = get_config_map_ref();
    // 将unordered_map转换为YAML::Node进行查找
    YAML::Node root;
    for (const auto& pair : config_map) {
        root[pair.first] = pair.second;
    }
    YAML::Node result = lookup_map(root, path);
    if (result.IsDefined()) {
        // 转换为std::any, 这里简化, 返回YAML::Node的字符串表示或其他
        // 实际上, 需要根据类型转换
        if (result.IsScalar()) {
            return result.as<std::string>();
        } else if (result.IsMap()) {
            // 返回整个子map, 但这里简化
            return result;
        }
        // 其他类型类似
    }
    return std::any();
}

// DecodeTo 将源数据解码到目标结构体
bool decode_to(void* dst, const std::any& src) {
    // 这里简化实现, 假设src是YAML::Node或string
    // 实际需要根据dst类型进行反序列化
    // 使用yaml-cpp进行转换
    try {
        if (src.type() == typeid(YAML::Node)) {
            YAML::Node node = std::any_cast<YAML::Node>(src);
            // 假设dst是指向BaseConfig或其他结构体的指针
            // 这里需要模板或类型检查, 但C++没有反射
            // 简化: 假设是BaseConfig
            BaseConfig* config = static_cast<BaseConfig*>(dst);
            *config = node.as<BaseConfig>();
            return true;
        } else if (src.type() == typeid(std::string)) {
            std::string str = std::any_cast<std::string>(src);
            YAML::Node node = YAML::Load(str);
            BaseConfig* config = static_cast<BaseConfig*>(dst);
            *config = node.as<BaseConfig>();
            return true;
        }
    } catch (const YAML::Exception& e) {
        // 错误处理
        return false;
    }
    return false;
}

} // namespace core
} // namespace quant1x